core-bundle/src/Resources/contao/library/Contao/Controller.php

```diff
@@ -1512,6 +1518,224 @@ protected function eliminateNestedPages($arrPages, $strTable=null, $blnSorting=f
+				'addBefore' => true,
+				'fullsize' => false,
+			);
+
+			if (null !== $filesModel)
+			{
+				// Set empty meta data (this is actually a minor BC break)
+				$templateData = array_replace_recursive($templateData, array(
+					'alt' => '',
+					'caption' => '',
+					'imageTitle' => '',
+					'imageUrl' => '',
+				));
+			}
+
+			return $templateData;
+		};
+
+		// Helper: Get size and margins and handle legacy $maxWidth option
+		$getSizeAndMargin = static function () use ($rowData, $maxWidth)
+		{
+			$size = $rowData['size'] ?? null;
+			$margin = StringUtil::deserialize($rowData['imagemargin'] ?? null);
+			$maxWidth = (int) ($maxWidth ?? Config::get('maxImageWidth'));
+
+			if (0 === $maxWidth)
+			{
+				return array($size, $margin);
+			}
+
+			@trigger_error('Using a maximum front end width has been deprecated and will no longer work in Contao 5.0. Remove the "maxImageWidth" configuration and use responsive images instead.', E_USER_DEPRECATED);
+
+			// Adjust margins if needed
+			if ('px' === ($margin['unit'] ?? null))
+			{
+				$horizontalMargin = (int) ($margin['left'] ?? 0) + (int) ($margin['right'] ?? 0);
+
+				if ($maxWidth - $horizontalMargin < 1)
+				{
+					$margin['left'] = '';
+					$margin['right'] = '';
+				}
+				else
+				{
+					$maxWidth -= $horizontalMargin;
+				}
+			}
+
+			// Normalize size
+			if ($size instanceof PictureConfiguration)
+			{
+				return array($size, $margin);
+			}
+
+			if (is_numeric($size))
+			{
+				$size = array(0, 0, (int) $size);
+			}
+			else
+			{
+				$size = (\is_array($size) ? $size : array()) + array(0, 0, 'crop');
+				$size[0] = (int) $size[0];
+				$size[1] = (int) $size[1];
+			}
+
+			// Adjust image size configuration if it exceeds the max width
+			if ($size[0] > 0 && $size[1] > 0)
+			{
+				list($width, $height) = $size;
+			}
+			else
+			{
+				$container = System::getContainer();
+
+				/** @var BoxInterface $originalSize */
+				$originalSize = $container
+					->get('contao.image.image_factory')
+					->create($container->getParameter('kernel.project_dir') . '/' . $rowData['singleSRC'])
+					->getDimensions()
+					->getSize();
+
+				$width = $originalSize->getWidth();
+				$height = $originalSize->getHeight();
+			}
+
+			if ($width <= $maxWidth)
+			{
+				return array($size, $margin);
+			}
+
+			$size[0] = $maxWidth;
+			$size[1] = floor($maxWidth * ($height / $width));
+
+			return array($size, $margin);
+		};
+
+		/** @var FigureBuilder $figureBuilder */
+		$figureBuilder = System::getContainer()->get(Studio::class)->createFigureBuilder();
+
+		// Set image resource
+		try
+		{
+			if (null !== $filesModel)
+			{
+				// Make sure model points to the same resource (BC)
+				$filesModel = clone $filesModel;
+				$filesModel->path = $rowData['singleSRC'];
+
+				// Use source + meta data from files model (if not overwritten)
+				$figureBuilder
+					->fromFilesModel($filesModel)
+					->setMetaData($createMetaDataOverwriteFromRowData(true));
```

Comment:
```suggestion
				$figureBuilder->fromFilesModel($filesModel);

				if (!($metaData = $createMetaDataOverwriteFromRowData(true))->empty())
				{
					$figureBuilder->setMetaData($metaData);
				}
```
If the meta data from raw data is empty, we should not overwrite the one from the files model. I noticed this BC break using a gallery element with an image that has `link` set in the meta data.

Reply:
No, your code is right, the actual issue is here: https://github.com/contao/contao/pull/1753/files#r449885615